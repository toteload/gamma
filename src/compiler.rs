use crate::{
    ast::{NodeId, NodeIdGenerator},
    error::Error,
    ink_codegen::CodeGenerator,
    parser::Parser,
    semantics::{validate_semantics, SemanticContext},
    source_location::SourceSpan,
    string_interner::{StringInterner, Symbol},
    type_check::type_check,
    types::{Signedness, Type, TypeInterner, TypeToken},
};
use std::collections::HashMap;

pub use crate::ink_codegen::{Options};

pub struct Context {
    pub id_generator: NodeIdGenerator,

    pub symbols: StringInterner,
    pub type_tokens: TypeInterner,
    pub type_table: HashMap<Symbol, TypeToken>,

    pub spans: HashMap<NodeId, SourceSpan>,
    pub types: HashMap<NodeId, TypeToken>,
}

impl Context {
    pub fn new() -> Self {
        let mut symbols = StringInterner::new();
        let mut type_tokens = TypeInterner::new();

        let type_table = HashMap::from([
            (
                symbols.add("i8"),
                type_tokens.add(Type::Int {
                    signedness: Signedness::Signed,
                    width: 8,
                }),
            ),
            (
                symbols.add("i16"),
                type_tokens.add(Type::Int {
                    signedness: Signedness::Signed,
                    width: 16,
                }),
            ),
            (
                symbols.add("i32"),
                type_tokens.add(Type::Int {
                    signedness: Signedness::Signed,
                    width: 32,
                }),
            ),
            (
                symbols.add("i64"),
                type_tokens.add(Type::Int {
                    signedness: Signedness::Signed,
                    width: 64,
                }),
            ),
            (
                symbols.add("u8"),
                type_tokens.add(Type::Int {
                    signedness: Signedness::Unsigned,
                    width: 8,
                }),
            ),
            (
                symbols.add("u16"),
                type_tokens.add(Type::Int {
                    signedness: Signedness::Unsigned,
                    width: 16,
                }),
            ),
            (
                symbols.add("u32"),
                type_tokens.add(Type::Int {
                    signedness: Signedness::Unsigned,
                    width: 32,
                }),
            ),
            (
                symbols.add("u64"),
                type_tokens.add(Type::Int {
                    signedness: Signedness::Unsigned,
                    width: 64,
                }),
            ),
            (symbols.add("void"), type_tokens.add(Type::Void)),
            (symbols.add("bool"), type_tokens.add(Type::Bool)),
        ]);

        Context {
            id_generator: NodeIdGenerator::new(),
            symbols,
            type_tokens,
            type_table,
            spans: HashMap::new(),
            types: HashMap::new(),
        }
    }

    pub fn compile(&mut self, source: &str, options: &Options) -> Result<String, Vec<Error>> {
        let mut parser = Parser::new(
            source,
            &mut self.symbols,
            &mut self.id_generator,
            &mut self.spans,
        );

        let items = parser.parse_items().map_err(|e| vec![e])?;

        let semantic_context = SemanticContext {
            symbols: &self.symbols,
            spans: &self.spans,
        };

        validate_semantics(&semantic_context, &items)?;

        type_check(
            &items,
            &mut self.type_tokens,
            &mut self.types,
            &mut self.type_table,
        )?;

        let ctx = inkwell::context::Context::create();
        let mut codegen = CodeGenerator::new(&ctx, &self.symbols, &self.types, &self.type_tokens);
        let output = codegen.compile(&items, options).map_err(|e| vec![e])?;

        Ok(output)
    }
}
