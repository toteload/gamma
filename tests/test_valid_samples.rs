use gamma::compiler::{{Context, Options, OutputTarget}};
use insta::assert_snapshot;
use std::fs;

#[test]
fn basic_arithmetic() {
    let contents = fs::read_to_string("tests/valid_samples/basic_arithmetic.gamma").unwrap();

    let mut context = Context::new();
    let result = context.compile(&contents, &Options { optimize: false, output: OutputTarget::LlvmIr });

    let Ok(output) = result else {
        let Err(errors) = result else { unreachable!() };

        for e in &errors {
            e.print(&context);
        }

        panic!("Compilation of sample \"basic_arithmetic\" resulted in {} error(s)", errors.len()); 
    };

    assert_snapshot!("basic_arithmetic", output);
}

#[test]
fn basic_main() {
    let contents = fs::read_to_string("tests/valid_samples/basic_main.gamma").unwrap();

    let mut context = Context::new();
    let result = context.compile(&contents, &Options { optimize: false, output: OutputTarget::LlvmIr });

    let Ok(output) = result else {
        let Err(errors) = result else { unreachable!() };

        for e in &errors {
            e.print(&context);
        }

        panic!("Compilation of sample \"basic_main\" resulted in {} error(s)", errors.len()); 
    };

    assert_snapshot!("basic_main", output);
}

#[test]
fn nested_loop() {
    let contents = fs::read_to_string("tests/valid_samples/nested_loop.gamma").unwrap();

    let mut context = Context::new();
    let result = context.compile(&contents, &Options { optimize: false, output: OutputTarget::LlvmIr });

    let Ok(output) = result else {
        let Err(errors) = result else { unreachable!() };

        for e in &errors {
            e.print(&context);
        }

        panic!("Compilation of sample \"nested_loop\" resulted in {} error(s)", errors.len()); 
    };

    assert_snapshot!("nested_loop", output);
}

#[test]
fn sum_loop() {
    let contents = fs::read_to_string("tests/valid_samples/sum_loop.gamma").unwrap();

    let mut context = Context::new();
    let result = context.compile(&contents, &Options { optimize: false, output: OutputTarget::LlvmIr });

    let Ok(output) = result else {
        let Err(errors) = result else { unreachable!() };

        for e in &errors {
            e.print(&context);
        }

        panic!("Compilation of sample \"sum_loop\" resulted in {} error(s)", errors.len()); 
    };

    assert_snapshot!("sum_loop", output);
}
#[test]
fn basic_arithmetic_optimized_assembly() {
    let contents = fs::read_to_string("tests/valid_samples/basic_arithmetic.gamma").unwrap();

    let mut context = Context::new();
    let result = context.compile(&contents, &Options { optimize: true, output: OutputTarget::Assembly });

    let Ok(output) = result else {
        let Err(errors) = result else { unreachable!() };

        for e in &errors {
            e.print(&context);
        }

        panic!("Compilation of sample \"basic_arithmetic_optimized_assembly\" resulted in {} error(s)", errors.len()); 
    };

    assert_snapshot!("basic_arithmetic_optimized_assembly", output);
}

#[test]
fn basic_main_optimized_assembly() {
    let contents = fs::read_to_string("tests/valid_samples/basic_main.gamma").unwrap();

    let mut context = Context::new();
    let result = context.compile(&contents, &Options { optimize: true, output: OutputTarget::Assembly });

    let Ok(output) = result else {
        let Err(errors) = result else { unreachable!() };

        for e in &errors {
            e.print(&context);
        }

        panic!("Compilation of sample \"basic_main_optimized_assembly\" resulted in {} error(s)", errors.len()); 
    };

    assert_snapshot!("basic_main_optimized_assembly", output);
}

#[test]
fn nested_loop_optimized_assembly() {
    let contents = fs::read_to_string("tests/valid_samples/nested_loop.gamma").unwrap();

    let mut context = Context::new();
    let result = context.compile(&contents, &Options { optimize: true, output: OutputTarget::Assembly });

    let Ok(output) = result else {
        let Err(errors) = result else { unreachable!() };

        for e in &errors {
            e.print(&context);
        }

        panic!("Compilation of sample \"nested_loop_optimized_assembly\" resulted in {} error(s)", errors.len()); 
    };

    assert_snapshot!("nested_loop_optimized_assembly", output);
}

#[test]
fn sum_loop_optimized_assembly() {
    let contents = fs::read_to_string("tests/valid_samples/sum_loop.gamma").unwrap();

    let mut context = Context::new();
    let result = context.compile(&contents, &Options { optimize: true, output: OutputTarget::Assembly });

    let Ok(output) = result else {
        let Err(errors) = result else { unreachable!() };

        for e in &errors {
            e.print(&context);
        }

        panic!("Compilation of sample \"sum_loop_optimized_assembly\" resulted in {} error(s)", errors.len()); 
    };

    assert_snapshot!("sum_loop_optimized_assembly", output);
}

