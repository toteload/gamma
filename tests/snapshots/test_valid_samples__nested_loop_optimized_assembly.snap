---
source: tests/test_valid_samples.rs
expression: output
---
	.text
	.def	@feat.00;
	.scl	3;
	.type	0;
	.endef
	.globl	@feat.00
.set @feat.00, 0
	.file	"main"
	.def	main;
	.scl	2;
	.type	32;
	.endef
	.globl	main
	.p2align	4, 0x90
main:
.seh_proc main
	pushq	%rbp
	.seh_pushreg %rbp
	movq	%rsp, %rbp
	.seh_setframe %rbp, 0
	.seh_endprologue
	xorl	%ecx, %ecx
	jmp	.LBB0_1
	.p2align	4, 0x90
.LBB0_4:
	incq	%rcx
	movq	%rdx, %rsp
.LBB0_1:
	cmpq	$10, %rcx
	je	.LBB0_6
	movq	%rsp, %rdx
	movl	$16, %eax
	callq	__chkstk
	subq	%rax, %rsp
	movq	%rsp, %rax
	movq	$0, (%rax)
	cmpq	$10, (%rax)
	je	.LBB0_4
	.p2align	4, 0x90
.LBB0_5:
	incq	(%rax)
	cmpq	$10, (%rax)
	jne	.LBB0_5
	jmp	.LBB0_4
.LBB0_6:
	xorl	%eax, %eax
	movq	%rbp, %rsp
	popq	%rbp
	retq
	.seh_endproc


